boot/initramfs.go

```diff
@@ -0,0 +1,72 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2020 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package boot
+
+import (
+	"fmt"
+
+	"github.com/snapcore/snapd/bootloader"
+	"github.com/snapcore/snapd/snap"
+)
+
+// TODO:UC20: add install/recover mode mounts code here too?
+
+// InitramfsRunModeChooseSnapsToMount returns a map of the snap paths to mount
+// for the specified snap types.
+func InitramfsRunModeChooseSnapsToMount(
+	typs []snap.Type,
+	blDir string,
+	blOpts *bootloader.Options,
+	modeenv *Modeenv,
+) (map[snap.Type]snap.PlaceInfo, error) {
+	var sn snap.PlaceInfo
+	var err error
+	m := make(map[snap.Type]snap.PlaceInfo)
+	for _, typ := range typs {
+		// TODO: it would be nice to use an interface here since the methods are
+		// already organized this way...
+		switch typ {
+		case snap.TypeBase:
+			bs := &bootState20Base{}
+			fmt.Println("bs.modeenv from top level func is", modeenv)
+			bs.modeenv = modeenv
```

Comment:
can't this be done above in one go?

Reply:
no, because go is weird about assigning to promoted members in a struct literal, see:

```
$ go build ./boot/...
boot/initramfs.go:46:5: cannot use promoted field bootState20Modeenv.modeenv in struct literal of type bootState20Base
```